[SYCL][Doc] Update sycl_ext_oneapi_local_memory

Pennycook opened the pull request:
- Fix return type of multi_ptr; Group::address_space does not exist.
- Clarify the constructor behavior for array types.

gmlueck commented:
I think this looks good.  Do you mind updating to the current [template](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc) also?  At a minimum, the Status section shoul dbe updated with the correct wording for a "supported" extension.

Pennycook commented:
> I think this looks good. Do you mind updating to the current [template](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc) also? At a minimum, the Status section shoul dbe updated with the correct wording for a "supported" extension.

Good idea.

I updated to the template in 9b49b85, and since I was editing things anyway I updated the descriptions as well in 6dc6182.

gmlueck commented on sycl/doc/extensions/supported/sycl_ext_oneapi_local_memory.asciidoc:
```diff
@@ -195,25 +194,7 @@ The example above creates a kernel with four work-groups, each containing 32
 work-items. An `int[64]` object is defined in group-local memory, and
 each work-item in the work-group obtains a `multi_ptr` to the same allocation.
```
I think it would be useful to explain here that the elements of the array initialized to zero, which might not be obvious to an average C++ user.  (The array is aggregate initialized, but there are not any `args` in the initializer list.  This causes the array elements to be value initialized, which means they are set to zero.)

gmlueck replied:
Thanks, that seems good.

Pennycook commented on sycl/doc/extensions/supported/sycl_ext_oneapi_local_memory.asciidoc:
```diff
@@ -126,55 +116,64 @@ work-items in the group, and is shared by all work-items of the group.
-initialized upon or before the first call to `group_local_memory`.  The storage
+_Effects_: Constructs an object of type `T` in an address space accessible by
+all work-items in group `g`, using default initialization.  The object is
+initialized upon or before the first call to `group_local_memory`. The storage
```
Fixed in 7b18e12.

Pennycook commented on sycl/doc/extensions/supported/sycl_ext_oneapi_local_memory.asciidoc:
```diff
@@ -126,55 +116,64 @@ work-items in the group, and is shared by all work-items of the group.
+all work-items in group `g`, using default initialization.  The object is
+initialized upon or before the first call to `group_local_memory`. The storage
 for the object is allocated upon or before the first call to
 `group_local_memory`, and deallocated when all work-items in the group have
```
Fixed in 7b18e12.

AlexeySachkov commented:
We seem to have implemented the change in #10367, but this PR is still in draft state. @Pennycook, are there any blockers preventing us to move further with it?

Pennycook commented:
> We seem to have implemented the change in #10367, but this PR is still in draft state. @Pennycook, are there any blockers preventing us to move further with it?

No, I think this is good to merge.  I originally had this as a draft because I expected the implementation might be appended to it.
